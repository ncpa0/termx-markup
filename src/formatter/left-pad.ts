export const leftPad = (str: string, length: number, char = " ") => {
  char = char[0] ?? " ";
  const pad = char.repeat(length);
  const lines = str.split("\n");
  return lines.map((line) => `${pad}${line}`).join("\n");
};
