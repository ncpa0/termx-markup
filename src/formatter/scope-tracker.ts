export type Scope = {
  parentTag?: string;
  color?: string;
  bg?: string;
  bold?: boolean;
  underscore?: boolean;
  strikethrough?: boolean;
  italic?: boolean;
  blink?: boolean;
  dimmed?: boolean;
  inverted?: boolean;
  noInherit?: boolean;
};

export class ScopeTracker {
  private static scopeStack: Scope[] = [
    {
      parentTag: "",
    },
  ];
  private static _currentScope: Scope = this.scopeStack[0]!;

  static get currentScope(): Scope {
    return this._currentScope;
  }

  static enterScope(scope: Scope) {
    const s = scope;
    this.scopeStack.push(s);
    this._currentScope = s;
  }

  static exitScope() {
    this.scopeStack.pop();
    this._currentScope = this.scopeStack[this.scopeStack.length - 1] ?? {};
  }
}
