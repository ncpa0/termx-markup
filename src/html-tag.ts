/**
 * Creates a html string from the given template literal. Each parameter
 * in the template literal is escaped to not include xml tag characters.
 *
 * @example
 *  html`<div>${"<span>hello</span>"}</div>`;
 *  // > <div>&lt;span&gt;hello&lt;/span&gt;</div>
 */
export function html(...args: any[]): string {
  const b = args[0];
  let c = "",
    a = 0,
    d = 0;
  for (c = b[0], a = 1, d = args.length; a < d; a++) {
    if (
      typeof args[a] === "object" &&
      args[a] !== null &&
      args[a].name === "RawHtml"
    ) {
      // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
      c += args[a].toString() + b[a];
    } else {
      // eslint-disable-next-line @typescript-eslint/restrict-plus-operands
      c += sanitizeHtml(args[a].toString()) + b[a];
    }
  }
  return c;
}

/**
 * Creates a RawHtml object that can be used to insert raw HTML
 * into a `html` template.
 *
 * @example
 *   html`<div>${"<span>hello</span>"}</div>`;
 *   // > <div>&lt;span&gt;hello&lt;/span&gt;</div>
 *
 *   html`<div>${raw("<span>hello</span>")}</div>`;
 *   // > <div><span>hello</span></div>
 */
export function raw(html: string): RawHtml {
  return new RawHtml(html);
}

function sanitizeHtml(html: string): string {
  return html.replace(/</g, "&lt;").replace(/>/g, "&gt;");
}

export function desanitizeHtml(html: string): string {
  return html.replace(/&lt;/g, "<").replace(/&gt;/g, ">");
}

class RawHtml {
  name = "RawHtml";
  constructor(public html: string) {}

  toString(): string {
    return this.html;
  }
}
