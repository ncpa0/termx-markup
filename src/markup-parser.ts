export type MarkupNode = {
  tag: string;
  textNode: boolean;
  attributes: Array<[attributeName: string, value: string | boolean]>;
  content: Array<string | MarkupNode>;
};

export class MarkupParserError extends Error {
  constructor(
    public parsedMarkup: string,
    public position: number,
    message: string
  ) {
    super("Invalid Markup. " + message);
  }

  private getPositionCords(): [line: number, column: number] {
    let lineIndex = 0;
    let columnIndex = 0;
    for (let i = 0; i < this.position; i++) {
      if (this.parsedMarkup[i] === "\n") {
        lineIndex++;
        columnIndex = 0;
      } else {
        columnIndex++;
      }
    }
    return [lineIndex, columnIndex];
  }

  private getStack() {
    return this.stack ?? "";
  }

  getPositionPatch() {
    const [lineIndex, columnIndex] = this.getPositionCords();

    const lines = this.parsedMarkup.split("\n");

    const patchLine = lines[lineIndex]!;

    const highlightLine = " ".repeat(columnIndex) + "^";

    let result = "";

    if (lineIndex > 1) {
      result += lines[lineIndex - 2]! + "\n";
    }

    if (lineIndex > 0) {
      result += lines[lineIndex - 1]! + "\n";
    }

    result += patchLine + "\n" + highlightLine;

    return result;
  }

  toString() {
    return `${
      this.message
    }\n\n${this.getPositionPatch()}\n\n${this.getStack()}`;
  }
}

class MarkupNodeBuilder {
  o: MarkupNode = {
    tag: "",
    content: [],
    attributes: [],
    textNode: false,
  };

  content: Array<string | MarkupNodeBuilder> = [];
  lastStringContent = "";

  constructor(public parent?: MarkupNodeBuilder) {}

  addChild() {
    if (this.lastStringContent.length) {
      this.content.push(this.lastStringContent);
      this.lastStringContent = "";
    }
    const child = new MarkupNodeBuilder(this);
    this.content.push(child);
    return child;
  }

  addChar(char: string) {
    this.lastStringContent += char;
  }

  serialize(): MarkupNode {
    for (let i = 0; i < this.content.length; i++) {
      const c = this.content[i]!;
      if (typeof c === "string") {
        this.o.content.push(c);
      } else {
        this.o.content.push(c.serialize());
      }
    }

    if (this.lastStringContent.length) {
      this.o.content.push(this.lastStringContent);
    }

    return this.o;
  }
}

class MarkupBuilder {
  /** MArkup node that is currently being built. */
  node: MarkupNodeBuilder = new MarkupNodeBuilder();

  constructor() {
    this.node.o.textNode = true;
  }

  addChildNode() {
    this.node = this.node.addChild();
  }

  moveUp() {
    this.node = this.node.parent!;
  }

  isTopLevel() {
    return this.node.parent === undefined;
  }

  serialize(): MarkupNode {
    if (
      this.node.content.length === 1 &&
      this.node.o.tag === "" &&
      typeof this.node.content[0] !== "string"
    ) {
      return this.node.content[0]!.serialize();
    }
    return this.node.serialize();
  }
}

const IS_IN_TAG = 1;
const IS_IN_ATTRIBUTE = 2;
const IS_IN_ATTRIBUTE_QUOTE = 4;
const IS_ESCAPED = 8;
const IS_CLOSING_TAG = 16;
const IS_TAG_NAME_READ = 32;

const NOT_IS_IN_TAG = ~IS_IN_TAG;
const NOT_IS_IN_ATTRIBUTE = ~IS_IN_ATTRIBUTE;
const NOT_IS_IN_ATTRIBUTE_QUOTE = ~IS_IN_ATTRIBUTE_QUOTE;
const NOT_IS_ESCAPED = ~IS_ESCAPED;
const NOT_IS_CLOSING_TAG = ~IS_CLOSING_TAG;
const NOT_IS_TAG_NAME_READ = ~IS_TAG_NAME_READ;

/**
 * Markup parsing function used internally by the library.
 *
 * It is only capable of parsing simple Markup structures and has
 * no support for most of the XML features like namespaces,
 * comments, CDATA, etc. But in return it is extremely fast. (up
 * to 3x faster than the
 * [fast-xml-parser](https://www.npmjs.com/package/fast-xml-parser))
 */
export function parseMarkup(markupStr: string): MarkupNode {
  const markup = new MarkupBuilder();

  let currentAttributeName = "";
  let currentAttributeValue = "";

  let closeTagName = "";

  let state = 0;

  for (let i = 0; i < markupStr.length; i++) {
    const char = markupStr[i]!;

    switch (state) {
      case 0: {
        switch (char) {
          case "<": {
            state = state | IS_IN_TAG;

            if (markupStr[i + 1] === "/") {
              state = state | IS_CLOSING_TAG;
              i++;
            } else {
              markup.addChildNode();
            }
            continue;
          }
          case "\\": {
            state = state | IS_ESCAPED;
            continue;
          }
        }
        markup.node.addChar(char);
        break;
      }
      // IS_ESCAPED
      case 8: {
        markup.node.addChar(char);
        state = state & NOT_IS_ESCAPED;
        break;
      }
      // IS_IN_TAG
      case 1: {
        switch (char) {
          case "\n": {
            if (markup.node.o.tag.length > 0) {
              state = state | IS_TAG_NAME_READ;
            }
            continue;
          }
          case " ": {
            if (markup.node.o.tag.length > 0) {
              state = state | IS_TAG_NAME_READ;
            }
            continue;
          }
          case ">": {
            if (markup.node.o.tag.length === 0) {
              throw new MarkupParserError(markupStr, i, "No tag name found.");
            }
            state = state & NOT_IS_IN_TAG;
            continue;
          }
          case "/": {
            if (markupStr[i + 1] === ">") {
              state = state & NOT_IS_IN_TAG;
              i++;
              markup.moveUp();
            } else {
              throw new MarkupParserError(
                markupStr,
                i,
                "Invalid character encountered."
              );
            }
            continue;
          }
          case "=": {
            // = char means this is actually an attribute, not a tag name
            // but since we are in this case, it means the tag name is empty
            throw new MarkupParserError(
              markupStr,
              i - markup.node.o.tag.length,
              "No tag name found."
            );
          }
        }
        markup.node.o.tag += char;
        break;
      }
      // IS_IN_TAG | IS_TAG_NAME_READ
      case 33: {
        switch (char) {
          case "\n": {
            continue;
          }
          case " ": {
            continue;
          }
          case ">": {
            state = state & NOT_IS_IN_TAG & NOT_IS_TAG_NAME_READ;
            continue;
          }
          case "/": {
            if (markupStr[i + 1] === ">") {
              state = state & NOT_IS_IN_TAG & NOT_IS_TAG_NAME_READ;
              i++;
              markup.moveUp();
            } else {
              throw new MarkupParserError(
                markupStr,
                i,
                "Invalid character encountered."
              );
            }
            continue;
          }
          case "=": {
            throw new MarkupParserError(
              markupStr,
              i,
              "Invalid character encountered."
            );
          }
        }
        state = state | IS_IN_ATTRIBUTE;
        currentAttributeName += char;
        break;
      }
      // IS_IN_TAG | IS_CLOSING_TAG
      case 17: {
        switch (char) {
          case "\n": {
            continue;
          }
          case " ": {
            continue;
          }
          case ">": {
            state = state & NOT_IS_IN_TAG & NOT_IS_CLOSING_TAG;
            if (closeTagName !== markup.node.o.tag) {
              throw new MarkupParserError(
                markupStr,
                i,
                `Closing tag does not match opening tag, expected '${markup.node.o.tag}' but found '${closeTagName}'.`
              );
            }
            closeTagName = "";
            markup.moveUp();
            continue;
          }
        }
        closeTagName += char;
        break;
      }
      // IS_IN_TAG | IS_TAG_NAME_READ | IS_IN_ATTRIBUTE
      case 35: {
        switch (char) {
          case "=": {
            state = state & NOT_IS_IN_ATTRIBUTE;
            if (markupStr[i + 1] === '"') {
              state = state | IS_IN_ATTRIBUTE_QUOTE;
              i += 1;
            } else {
              throw new MarkupParserError(
                markupStr,
                i + 1,
                "Attribute values must be enclosed in double quotes."
              );
            }
            continue;
          }
          case "\n": {
            state = state & NOT_IS_IN_ATTRIBUTE;
            markup.node.o.attributes.push([currentAttributeName, true]);
            currentAttributeName = "";
            continue;
          }
          case " ": {
            state = state & NOT_IS_IN_ATTRIBUTE;
            markup.node.o.attributes.push([currentAttributeName, true]);
            currentAttributeName = "";
            continue;
          }
          case ">": {
            state =
              state &
              NOT_IS_IN_TAG &
              NOT_IS_IN_ATTRIBUTE &
              NOT_IS_TAG_NAME_READ;
            markup.node.o.attributes.push([currentAttributeName, true]);
            currentAttributeName = "";
            continue;
          }
        }
        currentAttributeName += char;
        break;
      }
      // IS_IN_TAG | IS_TAG_NAME_READ | IS_IN_ATTRIBUTE_QUOTE
      case 37: {
        switch (char) {
          case '"': {
            state = state & NOT_IS_IN_ATTRIBUTE_QUOTE;
            markup.node.o.attributes.push([
              currentAttributeName,
              currentAttributeValue,
            ]);
            currentAttributeName = "";
            currentAttributeValue = "";
            continue;
          }
          case "\\": {
            state = state | IS_ESCAPED;
            continue;
          }
        }
        currentAttributeValue += char;
        break;
      }
      // IS_IN_TAG | IS_TAG_NAME_READ | IS_IN_ATTRIBUTE_QUOTE | IS_ESCAPED
      case 45: {
        currentAttributeValue += char;
        state = state & NOT_IS_ESCAPED;
        break;
      }
    }
  }

  if (!markup.isTopLevel()) {
    throw new MarkupParserError(
      markupStr,
      markupStr.length - 1,
      `Closing tag is missing. Expected a close tag for '${markup.node.o.tag}' before the end of the document.`
    );
  }

  return markup.serialize();
}
