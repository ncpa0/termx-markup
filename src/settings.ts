import type { Attributes } from "./formatter/scope-tracker";
import { DEFAULT_SCOPE } from "./formatter/scope-tracker";

export let GlobalSettings: Settings;

export class Settings {
  private static _instance: Settings;

  static {
    this._instance = new Settings();
    GlobalSettings = this._instance;
  }

  static setDefaultAttribute<K extends keyof Attributes>(
    attribute: K,
    value: Attributes[K]
  ) {
    this._instance.setDefaultAttribute(attribute, value);
  }

  static disableWarnings(disable: boolean) {
    this._instance.disableWarnings(disable);
  }

  /**
   * When strict mode is enabled, formatter will throw errors
   * instead of logging warnings.
   */
  static enableStrictMode(enable: boolean) {
    this._instance.enableStrictMode(enable);
  }

  private _settings = {
    warnings: true,
    strictMode: false,
  };

  setDefaultAttribute<K extends keyof Attributes>(
    attribute: K,
    value: Attributes[K]
  ) {
    DEFAULT_SCOPE.attributes[attribute] = value;
  }

  disableWarnings(disable: boolean) {
    this._settings.warnings = !disable;
  }

  /**
   * When strict mode is enabled, formatter will throw errors
   * instead of logging warnings.
   */
  enableStrictMode(enable: boolean) {
    this._settings.strictMode = enable;
  }

  /** @internal */
  get __() {
    return this._settings;
  }
}
