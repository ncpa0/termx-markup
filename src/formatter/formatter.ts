import { TermxBgColor } from "../colors/termx-bg-color";
import { TermxFontColor } from "../colors/termx-font-colors";
import { desanitizeHtml } from "../html-tag";
import type { MarkupNode } from "../markup-parser";
import { parseMarkup } from "../markup-parser";
import type { Scope } from "./scope-tracker";
import { ScopeTracker } from "./scope-tracker";

const escape = "\u001b";
const Bold = `${escape}[1m`;
const Dimmed = `${escape}[2m`;
const Italic = `${escape}[3m`;
const Underscore = `${escape}[4m`;
const Blink = `${escape}[5m`;
const Inverted = `${escape}[7m`;
const StrikeThrough = `${escape}[9m`;

/**
 * Formats given Markup into a string that can be printed to the
 * terminal.
 *
 * @example
 *   ```ts
 *   const markup = html`<p color="red">Hello World!</p>`;
 *
 *   const formatted = MarkupFormatter.format(markup);
 *   // formatted = "\u001b[31mHello World!\u001b[0m"
 *   ```
 */
export class MarkupFormatter {
  /**
   * Defines a new named color that can be used in the markup for
   * font and background colors.
   *
   * The given color can be in one of the following formats:
   *
   * - `#rrggbb` (hexadecimal)
   * - `rgb(r, g, b)` (decimal)
   * - `{ r: number, g: number, b: number }` (JSON object)
   */
  static defineColor(name: string, c: `#${string}` | `rgb(${string})`): void;
  static defineColor(
    name: string,
    c: { r: number; g: number; b: number }
  ): void;
  static defineColor(name: string, r: number, g: number, b: number): void;
  static defineColor(name: string, ...args: any[]) {
    TermxBgColor.define(name, ...args);
    TermxFontColor.define(name, ...args);
  }

  static format(markup: string): string {
    const node = parseMarkup(markup);
    return (
      TermxFontColor.get("unset") + desanitizeHtml(this.formatMarkup(node))
    );
  }

  private static formatMarkup(node: MarkupNode): string {
    let result = "";

    switch (node.tag) {
      case "pre":
      case "line":
      case "span": {
        ScopeTracker.enterScope(this.createScope(node));

        result +=
          this.scopeToTermMarks(ScopeTracker.currentScope) +
          this.join(
            node.content.map((content) =>
              this.mapContents(content, node.tag === "pre")
            )
          );

        ScopeTracker.exitScope();

        if (node.tag === "line") {
          result += "\n";
        }

        result +=
          TermxFontColor.get("unset") +
          this.scopeToTermMarks(ScopeTracker.currentScope);

        return result;
      }
      case "br": {
        result += "\n";
        return result;
      }
      case "": {
        result += this.join(
          node.content.map((content) => this.mapContents(content))
        );
        return result;
      }
    }

    throw new Error(`Invalid tag: <${node.tag}>`);
  }

  private static join(strings: string[]) {
    let result = "";
    for (let i = 0; i < strings.length; i++) {
      result += strings[i];
    }
    return result;
  }

  private static mapContents(content: MarkupNode | string, pre?: boolean) {
    if (typeof content === "string") {
      if (pre) {
        return content;
      }
      return content.replaceAll("\n", "").trim();
    }

    return this.formatMarkup(content);
  }

  private static scopeToTermMarks(scope: Scope): string {
    let result = "";

    if (scope.noInherit) {
      result += TermxFontColor.get("unset");
    }

    if (scope.color) {
      result += TermxFontColor.get(scope.color);
    }

    if (scope.bg) {
      result += TermxBgColor.get(scope.bg);
    }

    if (scope.bold) {
      result += Bold;
    }

    if (scope.dimmed) {
      result += Dimmed;
    }

    if (scope.italic) {
      result += Italic;
    }

    if (scope.underscore) {
      result += Underscore;
    }

    if (scope.blink) {
      result += Blink;
    }

    if (scope.inverted) {
      result += Inverted;
    }

    if (scope.strikethrough) {
      result += StrikeThrough;
    }

    return result;
  }

  private static createScope(node: MarkupNode): Scope {
    const noInherit = node.attributes.some(
      ([key, value]) =>
        key === "no-inherit" && (value === true || value === "true")
    );

    const scope: Scope = noInherit
      ? { noInherit: true }
      : { ...ScopeTracker.currentScope, noInherit: false };

    if (node.tag) {
      scope.parentTag = node.tag;
    }

    for (const [name, value] of node.attributes) {
      switch (name) {
        case "color":
          scope.color = as(value, "string");
          break;
        case "bg":
          scope.bg = as(value, "string");
          break;
        case "bold":
          scope.bold = value === true || value === "true";
          break;
        case "dim":
          scope.dimmed = value === true || value === "true";
          break;
        case "italic":
          scope.italic = value === true || value === "true";
          break;
        case "underscore":
          scope.underscore = value === true || value === "true";
          break;
        case "blink":
          scope.blink = value === true || value === "true";
          break;
        case "invert":
          scope.inverted = value === true || value === "true";
          break;
        case "strike":
          scope.strikethrough = value === true || value === "true";
          break;
      }
    }

    return scope;
  }
}

function as(value: string | boolean, as: "string"): string;
function as(value: string | boolean, as: "boolean"): boolean;
function as(value: string | boolean, as: "string" | "boolean") {
  if (typeof value === as) {
    return value;
  }
  throw new Error(`Invalid attribute type: ${typeof value} (expected ${as})`);
}
