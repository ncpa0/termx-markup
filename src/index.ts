import { MarkupFormatter } from "./formatter/formatter";
import { html, raw } from "./html-tag";
import { parseMarkup } from "./markup-parser";
import { Output, OutputBuffer } from "./output";
import { Settings } from "./settings";

export type { MarkupNode } from "./markup-parser";
export {
  MarkupFormatter,
  Output,
  OutputBuffer,
  Settings,
  html,
  parseMarkup,
  raw,
};

export default {
  MarkupFormatter,
  html,
  raw,
  Output,
  OutputBuffer,
  parseMarkup,
  Settings,
};
