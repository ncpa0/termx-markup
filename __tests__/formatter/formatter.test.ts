import { MarkupFormatter } from "../../src/formatter/formatter";
import { html, raw } from "../../src/html-tag";

describe("MarkupFormatter", () => {
  describe("should correctly format the xml", () => {
    it("scenario 1", () => {
      const xml = html`
        <line>Hello World</line>
        <line color="red">
          Red
          <pre color="blue"> or blue? </pre>
          text
        </line>
        <line>
          <span>This</span>
          <pre> is </pre>
          <pre>one </pre>
          <span>line</span>
        </line>
      `;

      const formatted = MarkupFormatter.format(xml);

      expect(formatted).toMatchSnapshot();
    });

    it("scenario 2", () => {
      const xml = html`
        <span>
          <span color="red"> Red </span>
          <span color="green"> Green </span>
          <span color="blue"> Blue </span>
          Normal
        </span>
      `;

      const formatted = MarkupFormatter.format(xml);

      expect(formatted).toMatchSnapshot();
    });

    it("scenario 3", () => {
      const xml = html`
        <span>
          <span color="red"> Red </span><br />
          <span color="green"> Green </span><br />
          <span color="blue"> Blue </span>< br /> Normal
        </span>
      `;

      const formatted = MarkupFormatter.format(xml);

      expect(formatted).toMatchSnapshot();
    });

    it("scenario 4", () => {
      const xml = html`
        <pre bold italic underscore>
          <span color="red"> Red </span>
          <span color="green"> Green </span>
        </pre>
      `;

      const formatted = MarkupFormatter.format(xml);

      expect(formatted).toMatchSnapshot();
    });

    it("scenario 5", () => {
      const xml = html`
        <span blink bold color="yellow">
          <pre>Lorem </pre>
          <span bold color="blue">
            <pre>ipsum </pre>
            <span dim color="green">
              <pre>dolor </pre>
              <span invert>
                <pre>sit </pre>
              </span>
              <pre>amet</pre>
            </span>
            <pre> consectetur </pre>
          </span>
          <pre>adipiscing </pre>
        </span>
      `;

      const formatted = MarkupFormatter.format(xml);

      expect(formatted).toMatchSnapshot();
    });

    it("scenario 6", () => {
      // prettier-ignore
      const xml = html`
        <span color="#f5aa42">
          <span color="rgb(137, 245, 209)"> Green </span>
          <pre> </pre>
          <span color="blue"> Blue </span>
          <pre> </pre>
          Orange
        </span>
      `;

      const formatted = MarkupFormatter.format(xml);

      expect(formatted).toMatchSnapshot();
    });

    it("scenario 7", () => {
      const xml = html`
        <span underscore color="yellow">
          <pre>Lorem </pre>
          <span no-inherit strike color="blue">
            <pre>ipsum </pre>
            <span dim color="green">
              <pre>dolor </pre>
            </span>
            <pre> sit </pre>
          </span>
          <pre>amet</pre>
        </span>
      `;

      const formatted = MarkupFormatter.format(xml);

      expect(formatted).toMatchSnapshot();
    });
  });

  describe("<repeat> tag", () => {
    it("should render as if it wasn't there if 'times' attribute is not specified", () => {
      const internalXml = html` <span color="red"> Red<s />Red </span>`;
      const xml = html` <repeat>${raw(internalXml)}</repeat> `;

      const formattedXml = MarkupFormatter.format(xml);

      expect(formattedXml).toMatchSnapshot();
    });

    it("should repeat it's content 2 times", () => {
      const internalXml = html`
        <span color="red">Red</span>
        <s />
        <span color="green">Green</span>
        <s />
        <span color="blue">Blue</span>
        <br />
      `;

      const xml = html` <repeat times="2">${raw(internalXml)}</repeat> `;

      const formattedXml = MarkupFormatter.format(xml);

      expect(formattedXml).toMatchSnapshot();
    });

    it("should repeat it's content 5 times", () => {
      const internalXml = html`
        <pre bold underscore bg="red" color="blue"> tf tf </pre>
        <br />
      `;

      const xml = html` <repeat times="5">${raw(internalXml)}</repeat> `;

      const formattedXml = MarkupFormatter.format(xml);

      expect(formattedXml).toMatchSnapshot();
    });

    it("should correctly repeat text content", () => {
      const xml = html`<repeat times="3">FOO</repeat>`;

      const formattedXml = MarkupFormatter.format(xml);

      expect(formattedXml).toMatchSnapshot();
    });
  });

  describe("<ol> tag", () => {
    it("should correctly render list with numbered indexes", () => {
      const xml = html`
        <ol>
          <li color="red">Red</li>
          <li color="green">Green</li>
          <li color="blue">Blue</li>
        </ol>
      `;

      const formattedXml = MarkupFormatter.format(xml);

      expect(formattedXml).toMatchSnapshot();
    });

    it("should correctly render nested list", () => {
      const xml = html`
        <ol>
          <li color="red">Red</li>
          <li color="green">Green</li>
          <li>
            <line>Shades of blue</line>
            <ol>
              <li color="blue">Blue</li>
              <li color="rgb(0, 147, 175)">Munsell</li>
              <li color="rgb(204, 204, 255)">Periwinkle</li>
            </ol>
          </li>
        </ol>
      `;

      const formattedXml = MarkupFormatter.format(xml);

      expect(formattedXml).toMatchSnapshot();
    });

    it("should correctly render multiline list elements", () => {
      const xml = html`
        <ol>
          <li color="red">
            <line>Red</line>
            <span>Red</span>
          </li>
          <li color="green">
            <line>Green</line>
            <span>Green</span>
          </li>
          <li color="blue">
            <line>Blue</line>
            <span>Blue</span>
          </li>
        </ol>
      `;

      const formattedXml = MarkupFormatter.format(xml);

      expect(formattedXml).toMatchSnapshot();
    });

    it("should correctly render multiline list elements with more than 9 elements", () => {
      const xml = html`
        <ol>
          <li color="red">
            <line>Red</line>
            <span>Red</span>
          </li>
          <li color="green">
            <line>Green</line>
            <span>Green</span>
          </li>
          <li color="blue">
            <line>Blue</line>
            <span>Blue</span>
          </li>
          <li color="yellow">
            <line>Yellow</line>
            <span>Yellow</span>
          </li>
          <li color="magenta">
            <line>Magenta</line>
            <span>Magenta</span>
          </li>
          <li color="cyan">
            <line>cyan</line>
            <span>cyan</span>
          </li>
          <li color="white">
            <line>white</line>
            <span>white</span>
          </li>
          <li color="lightRed">
            <line>lightRed</line>
            <span>lightRed</span>
          </li>
          <li color="lightGreen">
            <line>lightGreen</line>
            <span>lightGreen</span>
          </li>
          <li color="lightYellow">
            <line>lightYellow</line>
            <span>lightYellow</span>
          </li>
          <li color="lightBlue">
            <line>lightBlue</line>
            <span>lightBlue</span>
          </li>
        </ol>
      `;

      const formattedXml = MarkupFormatter.format(xml);

      expect(formattedXml).toMatchSnapshot();
    });

    it("should correctly render multiline list elements with more than 9 nested elements", () => {
      const xml = html`
        <ol>
          <li>Element 1</li>
          <li>Element 2</li>
          <li>
            <line>Colors:</line>
            <ol>
              <li color="red">
                <line>Red</line>
                <span>Red</span>
              </li>
              <li color="green">
                <line>Green</line>
                <span>Green</span>
              </li>
              <li color="blue">
                <line>Blue</line>
                <span>Blue</span>
              </li>
              <li color="yellow">
                <line>Yellow</line>
                <span>Yellow</span>
              </li>
              <li color="magenta">
                <line>Magenta</line>
                <span>Magenta</span>
              </li>
              <li color="cyan">
                <line>cyan</line>
                <span>cyan</span>
              </li>
              <li color="white">
                <line>white</line>
                <span>white</span>
              </li>
              <li color="lightRed">
                <line>lightRed</line>
                <span>lightRed</span>
              </li>
              <li color="lightGreen">
                <line>lightGreen</line>
                <span>lightGreen</span>
              </li>
              <li color="lightYellow">
                <line>lightYellow</line>
                <span>lightYellow</span>
              </li>
              <li color="lightBlue">
                <line>lightBlue</line>
                <span>lightBlue</span>
              </li>
              <li color="lightMagenta">
                <line>lightMagenta</line>
                <span>lightMagenta</span>
              </li>
            </ol>
          </li>
        </ol>
      `;

      const formattedXml = MarkupFormatter.format(xml);

      expect(formattedXml).toMatchSnapshot();
    });
  });

  describe("<ul> tag", () => {
    it("should correctly render list with bullet indexes", () => {
      const xml = html`
        <ul>
          <li color="red">Red</li>
          <li color="green">Green</li>
          <li color="blue">Blue</li>
        </ul>
      `;

      const formattedXml = MarkupFormatter.format(xml);

      expect(formattedXml).toMatchSnapshot();
    });

    it("should correctly render nested list", () => {
      const xml = html`
        <ul>
          <li color="red">Red</li>
          <li color="green">Green</li>
          <li>
            <line>Shades of blue</line>
            <ul type="circle">
              <li color="blue">Blue</li>
              <li color="rgb(0, 147, 175)">Munsell</li>
              <li color="rgb(204, 204, 255)">Periwinkle</li>
            </ul>
          </li>
        </ul>
      `;

      const formattedXml = MarkupFormatter.format(xml);

      expect(formattedXml).toMatchSnapshot();
    });

    it("should correctly render multiline list elements", () => {
      const xml = html`
        <ul type="square">
          <li color="red">
            <line>Red</line>
            <span>Red</span>
          </li>
          <li color="green">
            <line>Green</line>
            <span>Green</span>
          </li>
          <li color="blue">
            <line>Blue</line>
            <span>Blue</span>
          </li>
        </ul>
      `;

      const formattedXml = MarkupFormatter.format(xml);

      expect(formattedXml).toMatchSnapshot();
    });
  });
});
