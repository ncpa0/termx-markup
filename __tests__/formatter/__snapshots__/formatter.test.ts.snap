// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MarkupFormatter should correctly format the xml scenario 1 1`] = `
"[0mHello World
[0m[31mRed[34m or blue? [0m[31mtext
[0mThis[0m is [0mone [0mline[0m
[0m"
`;

exports[`MarkupFormatter should correctly format the xml scenario 2 1`] = `"[0m[31mRed[0m[32mGreen[0m[34mBlue[0mNormal[0m"`;

exports[`MarkupFormatter should correctly format the xml scenario 3 1`] = `
"[0m[31mRed[0m
[32mGreen[0m
[34mBlue[0m
Normal[0m"
`;

exports[`MarkupFormatter should correctly format the xml scenario 4 1`] = `
"[0m[1m[3m[4m
          [31m[1m[3m[4mRed[0m[1m[3m[4m
          [32m[1m[3m[4mGreen[0m[1m[3m[4m
        [0m"
`;

exports[`MarkupFormatter should correctly format the xml scenario 5 1`] = `"[0m[33m[1m[5m[33m[1m[5mLorem [0m[33m[1m[5m[34m[1m[5m[34m[1m[5mipsum [0m[34m[1m[5m[32m[1m[2m[5m[32m[1m[2m[5mdolor [0m[32m[1m[2m[5m[32m[1m[2m[5m[7m[32m[1m[2m[5m[7msit [0m[32m[1m[2m[5m[7m[0m[32m[1m[2m[5m[32m[1m[2m[5mamet[0m[32m[1m[2m[5m[0m[34m[1m[5m[34m[1m[5m consectetur [0m[34m[1m[5m[0m[33m[1m[5m[33m[1m[5madipiscing [0m[33m[1m[5m[0m"`;

exports[`MarkupFormatter should correctly format the xml scenario 6 1`] = `"[0m[38;2;245;170;66m[38;2;137;245;209mGreen[0m[38;2;245;170;66m[38;2;245;170;66m [0m[38;2;245;170;66m[34mBlue[0m[38;2;245;170;66m[38;2;245;170;66m [0m[38;2;245;170;66mOrange[0m"`;

exports[`MarkupFormatter should correctly format the xml scenario 7 1`] = `"[0m[33m[4m[33m[4mLorem [0m[33m[4m[0m[34m[9m[34m[9mipsum [0m[0m[34m[9m[32m[2m[9m[32m[2m[9mdolor [0m[32m[2m[9m[0m[0m[34m[9m[34m[9m sit [0m[0m[34m[9m[0m[33m[4m[33m[4mamet[0m[33m[4m[0m"`;
